src/richie/apps/courses/admin.py

```diff
@@ -25,6 +38,101 @@ def title(self, obj):
         """
         return obj.extended_object.get_title()
 
+    def get_urls(self):
+        """
+        Add admin URL to trigger the snapshot of a course.
+        """
+        url_patterns = super().get_urls()
+        return [
+            url(
+                r"^(?P<course_id>[0-9]+)/snapshot/$",
+                self.admin_site.admin_view(self.snapshot),
+                name="cms_course_snapshot",
+            )
+        ] + url_patterns
+
+    @REQUIRE_POST
+    @transaction.atomic
+    def snapshot(self, request, course_id, *args, **kwargs):
+        """
+        Snapshotting a course is making a copy of the course page with all its permissions and
+        extensions, and placing the copy as the first child of the page being snapshotted, then
+        moving all the course run of the page being snapshotted as children of the new snapshot
+        so we keep record of the course as it was when these course runs were played.
+        """
+        try:
+            page = Page.objects.select_related("node__site").get(
+                publisher_is_draft=True, course__id=course_id
+            )
+        except Page.DoesNotExist:
+            return jsonify_request(
+                HttpResponseBadRequest(
+                    force_text(_("Error! Course could not be found."))
+                )
+            )
+
+        # If the page has a parent that is a course page, it is a snaptshot and should therefore
+        # not be allowed to be itself snapshotted.
+        if page.parent_page:
+            try:
+                page.parent_page.course
+            except Course.DoesNotExist:
+                pass
+            else:
+                return jsonify_request(
+                    HttpResponseForbidden(
+                        force_text(_("Error! You can't snapshot a snapshot."))
```

Comment:
![2mg3vy](https://user-images.githubusercontent.com/1932937/48425845-ec7b0b00-e765-11e8-8ec6-f23ea92e1f6f.jpg)

Reply:
Hope I can merge without resolving this one...